lib/s3_inventory.rb

```diff
@@ -0,0 +1,195 @@
+require "aws-sdk-s3"
+require "csv"
+require "discourse_event"
+
+class S3Inventory
+
+  class StorageError < RuntimeError; end
+
+  ::DiscourseEvent.on(:site_setting_saved) do |site_setting|
+    name = site_setting.name.to_s
+    Jobs.enqueue(:update_s3_inventory) if name.include?("s3_inventory") || name == "s3_upload_bucket"
+  end
+
+  attr_reader :inventory_id, :s3_client, :csv_filename,
+              :source_bucket_name, :source_bucket_path,
+              :destination_bucket_name, :destination_bucket_path
+
+  CSV_ETAG_INDEX ||= 2.freeze
+
+  def initialize(inventory_id = "uploads")
+    @source_bucket_name, @source_bucket_path = begin
+      raise Discourse::InvalidParameters.new("s3_upload_bucket") if SiteSetting.s3_upload_bucket.blank?
+      S3Helper.get_bucket_and_folder_path(SiteSetting.s3_upload_bucket)
+    end
+
+    @destination_bucket_name, @destination_bucket_path = begin
+      raise Discourse::InvalidParameters.new("s3_inventory_bucket") if SiteSetting.s3_inventory_bucket.blank?
+      S3Helper.get_bucket_and_folder_path(SiteSetting.s3_inventory_bucket)
+    end
+
+    s3_options = S3Helper.s3_options(SiteSetting)
+    @s3_helper = S3Helper.new(SiteSetting.s3_inventory_bucket, '', s3_options)
+    @s3_client = @s3_helper.s3_client
+
+    @inventory_id = inventory_id
+  end
+
+  def file
+    @file ||= unsorted_files.sort_by { |file| -file.last_modified.to_i }.first
+  end
+
+  def list_missing_uploads(skip_optimized: false)
+    if file.blank?
+      Rails.logger.warn("Failed to list inventory from S3")
+      raise StorageError
+    end
+
+    @current_db = RailsMultisite::ConnectionManagement.current_db
+    @timestamp = Time.now.strftime("%Y-%m-%d-%H%M%S")
+    @tmp_directory = File.join(Rails.root, "tmp", "inventory", @current_db, @timestamp)
+    @archive_filename = File.join(@tmp_directory, File.basename(file.key))
+    @csv_filename = @archive_filename[0...-3]
+
+    FileUtils.mkdir_p(@tmp_directory)
+    copy_archive_to_tmp_directory
+    unzip_archive
+
+    etags = []
+    connection = ActiveRecord::Base.connection.raw_connection
+    connection.exec('CREATE TEMP TABLE etags(val text PRIMARY KEY)')
```

Comment:
What if two processes call this method at the same time? Would it fail because we can't create two temp tables of the same name? Also, what was the rational for using a temp table instead of holding the records in memory?

Reply:
> What if two processes call this method at the same time? Would it fail

I think so. I have to fix it.

> what was the rational for using a temp table instead of holding the records in memory?

The reason is to check missing objects [using simple `JOIN` query](https://github.com/discourse/discourse/pull/6867/files/6fa012dbf08131ac757ad2583a9f1c45ee402a14#diff-7fc755e5e9dc2118547f063b20bbb2ddR86).

Reply:
> The reason is to check missing objects using simple JOIN query.

This is quite a nice idea :+1:  Let's try it out and see how it works.